File: 0155-min-stack/0155-min-stack.js
class MinStack {
    constructor() {
        this.stack = [];
        this.minimum = Infinity;
    }

    push(value) {
        this.stack.push(value);
        this.minimum = Math.min(value, this.minimum);
    }

    pop() {
        const element = this.stack.pop();
        if (element === this.minimum) {
            this.minimum = Infinity;
            for (let i = this.stack.length - 1; i >= 0; i--) {
                this.minimum = Math.min(this.minimum, this.stack[i]);
            }
        } else {
            return this.minimum;
        }
    }

    getMin() {
        return this.minimum;
    }

    top() {
        return this.stack[this.stack.length - 1];
    }
}

File: 0155-min-stack/0155-min-stack.py
class MinStack:
    def __init__(self):
        self.stack = []
        self.minimum = float("inf")

    def push(self, value: int) -> None:
        self.minimum = min(self.minimum, value)
        self.stack.append(value)

    def pop(self) -> None:
        elem = self.stack.pop()
        if elem == self.minimum:
            self.minimum = float("inf")
            for i in range(len(self.stack)):
                self.minimum = min(self.minimum, self.stack[i])

    def top(self) -> int:
        if self.stack:
            return self.stack[-1]
        return 0

    def getMin(self) -> int:
        if self.minimum:
            return self.minimum
        return 0

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(value)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()